src/pgwire/server.rs

```diff
@@ -90,12 +100,78 @@ impl Server {
                     }
                     // For security, the client is not told whether the cancel
                     // request succeeds or fails.
-                    break Ok(());
+                    return Ok(());
                 }
 
-                FrontendStartupMessage::GssEncRequest => conn.write_all(&[b'N']).await?,
-                FrontendStartupMessage::SslRequest => conn.write_all(&[b'N']).await?,
+                FrontendStartupMessage::SslRequest => match conn {
+                    // NOTE(benesch): we can match on `self.tls` properly,
+                    // instead of checking `is_some` and `unwrap`ping, when
+                    // the move_ref_patterns feature stabilizes.
+                    // See: https://github.com/rust-lang/rust/issues/68354
+                    Conn::Unencrypted(mut conn) if self.tls.is_some() => {
+                        conn.write_all(&[b'S']).await?;
+                        let tls = self.tls.as_ref().unwrap();
+                        Conn::Ssl(tokio_openssl::accept(tls, conn).await?)
+                    }
+                    mut conn => {
+                        conn.write_all(&[b'N']).await?;
+                        conn
+                    }
+                },
+
+                FrontendStartupMessage::GssEncRequest => {
+                    conn.write_all(&[b'N']).await?;
```

Comment:
Future thought: I'm tempted to say we should add consts for all these bytes, it'd be easier to think about this review for fresh eyes if this said something like `conn.write_all(NOT_SUPPORTED).await?`

Reply:
👍 done